### DISCUSSION TITLE
Need ideas on how I should go about vehicle attached physics bodies

### DISCUSSION THREAD

**Discussion by Pindrought** [jrouwe/JoltPhysics#429] [Q&A] [answered]
I have my own custom 3d format that I process gltf files into.
I was going to use the V-HACD for performing convex decomposition to break up/bake my mesh into multiple convex hull shapes.
https://github.com/kmammou/v-hacd

For simpler objects like couches, tables, things with no moving parts it should be simple enough to just automatically generate the colliders when I convert the file from gltf->my custom format.

I've ran into a situation though regarding vehicles where i'd like to get other people's input on how they've gone about it.

For this example, I have a truck. 
![truck1](https://user-images.githubusercontent.com/39708313/216883704-85de6753-8249-414e-a36e-e7373d02c39f.png)

Since the wheels will be separate colliders from the body when I build my vehicle shape, I assume i'll remove them from the truck model.
![truck2](https://user-images.githubusercontent.com/39708313/216883720-dea09b37-c356-443e-ae9d-d3d3f8036b43.png)

So far, so good. However, now I see an issue. What if I want the option to lower the tailgate and have that reflected in my physics system?
![truck3](https://user-images.githubusercontent.com/39708313/216883866-7ac8ed72-1179-4081-abd2-275742f20d91.png)

If I am just automatically generating the whole shape, this is a problem since this would need to be it's own physics body.

Similarly, what if I want to have the door open and have collision with the door be a thing?
![truck4](https://user-images.githubusercontent.com/39708313/216884019-a2242b57-ed4e-42b5-ae98-6d0778c7e2a2.png)

Or what if I want the ability to remove the physics colliders for windows if they were destroyed?

Am I correct in assuming all of these parts that would need to move/possibly be removed should be their own bodies?

What type of constraints should I be using for connecting the door physicsbody to the vehicle body's physicsbody? **(Possibly powered hinge constraint?)** 
Or the tailgate to the vehicle body? **(Possibly powered hinge constraint?)** 
Or the window to the door? **(Not sure which one makes sense here? Maybe fixed constraint? However, fixed constraints seem like they still move around at the attachment point...)** 

I want to make sure that it's something that could be "locked" where it wouldn't be swinging around when i'm driving, but also have the ability to swing it open for the door/tailgates, and then for the windows i'd just want something locked in place until I remove that physics body entirely.

**Comment by jrouwe** [comment, marked as answer]
In general I would suggest:

* Export car body, tailgate, doors, windows separately and create a ConvexHullShape for each of them (or possibly a StaticCompoundShape of ConvexHullShapes if you use v-hacd)
* Combine those parts in a MutableCompoundShape in case you don't want them to move independently and use that shape as your car body
* When a part is destroyed or needs to start moving independently (e.g. door), remove the part from the MutableCompoundShape
* Create a new Dynamic body with just the shape of the part that you removed and place it exactly where it had been before
* Create a HingeConstraint to connect the two bodies (make sure you specify the attachment point correctly, i.e. where the hinge would be in real life, if you don't the hinge will be very sloppy)
* Either use the limits to limit the movement and/or set the motor state to 'Position' and configure the motor settings 'frequency' to get the desired spring strength
